src/v/cluster/tm_stm_types.cc

```diff
@@ -10,24 +10,99 @@
  */
 #include "cluster/tm_stm_types.h"
 
+#include <optional>
+
 namespace cluster {
 
+namespace {
+template<typename... Args>
+bool is_one_of(tx_status status, Args&&... args) {
+    return ((args == status) || ...);
+}
+
+} // namespace
+
+state_transition_error::state_transition_error(tx_status from, tx_status to)
+  : from(from)
+  , to(to) {}
+
+bool is_state_transition_valid(
+  const tx_metadata& current, tx_status requested_status) {
+    /**
+     * The state transition validation logic is really a simple definition of
+     * transaction fsm. The validation checks it the current state is a valid
+     * precursor of the requested one.
+     */
+    switch (requested_status) {
+    case empty:
+        // ready is an initial state a transaction can never go back to that
+        // state
+        return is_one_of(current.status, tx_status::empty);
+    case ongoing:
+        return is_one_of(
+          current.status,
+          tx_status::empty,
+          tx_status::ongoing,
+          tx_status::completed_abort,
+          tx_status::completed_commit);
```

Comment:
this transition is a bit non intuitive IMO, especially when we say empty is the begin state.

Reply:
true, i need to remove that transition